### ISSUE TITLE
Should we be using strerror_r?

### ISSUE THREAD

**Issue by jnthn** [MoarVM/MoarVM#1041]
Apparently [`strerror` ain't thread safe](http://man7.org/linux/man-pages/man3/strerror.3.html).

**Comment by Leont** [comment]
It seeems [UV](http://docs.libuv.org/en/v1.x/errors.html) provides `uv_strerror` and `uv_strerror_r`

**Comment by jnthn** [comment]
Ah, so we'd need to switch all those over too. That'll be quite the diff...

**Comment by Altai-man** [comment]
```
➜  MoarVM git:(master) rg -c 'strerror' # Includes also `uv_strerror`
src/moar.c:3
src/io/procops.c:5
src/io/asyncsocket.c:8
src/io/asyncsocketudp.c:5
src/io/filewatchers.c:1
src/io/syncsocket.c:3
src/io/syncfile.c:8
src/io/dirops.c:7
src/io/fileops.c:9
src/profiler/telemeh.c:1
src/profiler/heapsnapshot.c:1
src/core/fixedsizealloc.c:1
src/core/compunit.c:4
src/debug/debugserver.c:6
src/jit/dump.c:1
src/strings/nfg.c:1
src/core/threads.c:1
src/platform/win32/time.c:1
src/platform/posix/time.c:1
src/6model/reprs/ReentrantMutex.c:1
src/6model/reprs/ConditionVariable.c:1
src/6model/reprs/Semaphore.c:1
src/6model/reprs/ConcBlockingQueue.c:3
➜  MoarVM git:(master) 
```
This doesn't look to me like an awfully long diff to make, no?

If nobody opposes, I can do the change, recompile && run spectest && prepare a PR.

**Comment by pmqs** [comment]
Also, in **src/io/dirops.c** I see `readdir` is used. That isn't thread-safe either. Need to use `readdir_r`

**Comment by ugexe** [comment]
readdir not used in an thread unsafe way like strerror: https://colabti.org/irclogger/irclogger_log/moarvm?date=2019-03-29#l96

**Comment by MasterDuke17** [comment]
Does `readdir_r` provide any benefits aside from (the not-actually-needed-in-this-case) thread safety? If so, maybe it's worth changing to it. If not, maybe it's not worth the code churn.

**Comment by ugexe** [comment]
I doubt it considering libuv did not implemented its readdir ops using `readdir_r`.

https://www.gnu.org/software/libc/manual/html_node/Reading_002fClosing-Directory.html
> readdir_r allows you to provide your own buffer for the struct dirent, but it is less portable than readdir, and has problems with very long filenames (see below). We recommend you use readdir, but do not share DIR objects.

**Comment by pmqs** [comment]
> Does `readdir_r` provide any benefits aside from (the not-actually-needed-in-this-case) thread safety? 

Yes - it is reentrant. (see [here](https://deadbeef.me/2017/09/reentrant-threadsafe) for a nice discussion of the difference between multi-threaded & reentrant).

As a side not, in a previous multi-threaded project we took the decision to _always_ use the `r_`  versions of all these functions, even if they were being used in a content where it was unnecessary.  Meant that you couldn't accidentally add a non-thread safe or non-reentrant function.

**Comment by ugexe** [comment]
`readdir_r` should never recurse because it does not run any user code. It being reentrant is just a precursor to its requirement of being thread safe. So such a change would make a trade off of making auditing the code slightly easier at the cost of technical debt  (deprecated; not portable).
